swift/extractor/Extractor.cpp

```diff
@@ -0,0 +1,101 @@
+#include "Extractor.h"
+
+#include <filesystem>
+#include <fstream>
+#include <iostream>
+#include <sstream>
+#include <memory>
+#include <unistd.h>
+
+#include <swift/AST/SourceFile.h>
+#include <llvm/ADT/SmallString.h>
+#include <llvm/Support/FileSystem.h>
+#include <llvm/Support/Path.h>
+
+using namespace codeql;
+
+Extractor::Extractor(const Configuration& config, swift::CompilerInstance& instance)
+    : config{config}, compiler{instance} {}
+
+void Extractor::extract() {
+  if (compiler.getPrimarySourceFiles().empty()) {
+    swift::ModuleDecl* module = compiler.getMainModule();
+    if (!module->getFiles().empty() &&
+        module->getFiles().front()->getKind() == swift::FileUnitKind::Source) {
+      /// We can only call getMainSourceFile if the first file is of a Source kind
+      swift::SourceFile& file = module->getMainSourceFile();
+      extractFile(file);
+    }
+  } else {
+    for (auto s : compiler.getPrimarySourceFiles()) {
+      extractFile(*s);
+    }
+  }
+}
+
+void Extractor::extractFile(swift::SourceFile& file) {
+  if (std::error_code ec = llvm::sys::fs::create_directories(config.trapDir)) {
+    std::cerr << "Cannot create TRAP directory: " << ec.message() << "\n";
+    return;
+  }
+
+  if (std::error_code ec = llvm::sys::fs::create_directories(config.sourceArchiveDir)) {
+    std::cerr << "Cannot create source archive directory: " << ec.message() << "\n";
+    return;
+  }
+
+  llvm::SmallString<PATH_MAX> srcFilePath(file.getFilename());
```

Comment:
[llvm::SmallString docs](https://llvm.org/docs/ProgrammersManual.html#llvm-adt-smallstring-h).
Perhaps, we can use something smaller than `PATH_MAX` here, but any other number would be arbitrary. OTOH, QL db filenames can get rather long (especially on macOS).

Reply:
Might be good to intro a typedef, so you don't have to write `llvm::SmallString<PATH_MAX>` all over the place an think every time what the template parameter should be.

Reply:
I'll leave it as is for now and will revisit when we have more than one file/function using the SmallString for file/paths related tasks.